pkg/build/controller/build/build_controller.go

```diff
@@ -388,34 +409,65 @@ func (bc *BuildController) resolveOutputDockerImageReference(build *buildapi.Bui
 			var ok bool
 			streamName, tag, ok = imageapi.SplitImageStreamTag(streamName)
 			if !ok {
-				return "", fmt.Errorf("the referenced image stream tag is invalid: %s", outputTo.Name)
+				return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced image stream tag is invalid: %s", outputTo.Name)
 			}
 			tag = ":" + tag
 		}
 		stream, err := bc.imageStreamStore.ImageStreams(namespace).Get(streamName)
 		if err != nil {
 			if errors.IsNotFound(err) {
-				return "", fmt.Errorf("the referenced output image stream %s/%s does not exist", namespace, streamName)
+				return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced output image stream %s/%s does not exist", namespace, streamName)
 			}
-			return "", fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
+			return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
 		}
 		if len(stream.Status.DockerImageRepository) == 0 {
 			e := fmt.Errorf("the image stream %s/%s cannot be used as the output for build %s/%s because the integrated Docker registry is not configured and no external registry was defined", namespace, outputTo.Name, build.Namespace, build.Name)
 			bc.recorder.Eventf(build, kapi.EventTypeWarning, "invalidOutput", "Error starting build: %v", e)
-			return "", e
+			return "", build.Spec.Output.PushSecret, e
 		}
 		ref = fmt.Sprintf("%s%s", stream.Status.DockerImageRepository, tag)
 	}
-	return ref, nil
+
+	pushSecret = build.Spec.Output.PushSecret
+	if pushSecret == nil {
+		serviceAccount := build.Spec.ServiceAccount
+		if len(serviceAccount) == 0 {
```

Comment:
Seems like it would be cleaner to get this on your object in one spot (strategy?) rather that defaulting here and everywhere else.